staging/src/k8s.io/kubectl/pkg/cmd/expose/expose.go

```diff
@@ -378,3 +398,250 @@ func (o *ExposeServiceOptions) RunExpose(cmd *cobra.Command, args []string) erro
 	}
 	return nil
 }
+
+func generateService(genericParams map[string]interface{}) (runtime.Object, error) {
```

Comment:
Once you move the options to be part of `ExposeServiceOptions` (and considering we're not going to use the create.Run(), you should probably want to make this a method of the options:

```
func (o *ExposeServiceOptions) createService() (*corev1.Service, error) {
```

And use the fields from o. instead of the map 'params')

Reply:
type of `service` will be `*corev1.Service` so it will give a type mismatch at line `343` and also do I need to create `Override string` in `struct`  for `override` as it will be used at `341`

Reply:
1) Do you mean  `object, err = cmdutil.Merge(codec, object, inline)` ? The function takes in `runtime.Object` and `runtime.Object` is an `interface` (https://github.com/kubernetes/apimachinery/blob/v0.20.4/pkg/runtime/interfaces.go#L299) so I don't think there will be an type misnatch
2) Yes you will need to add `Override string`.

Reply:
1) Do you mean  `object, err = cmdutil.Merge(codec, object, inline)` ? The function takes in a `runtime.Object` and `runtime.Object` is an `interface` (https://github.com/kubernetes/apimachinery/blob/v0.20.4/pkg/runtime/interfaces.go#L299) so I don't think there will be an type mismatch
2) Yes you will need to add `Override string`.

Reply:
@lauchokyip as object will be renamed to service at `336` and also its type will be changed  other than `runtime.Object`
 so it was actually giving error stating that its type is different

Reply:
@almas33 line 336, do you mean by this line? `object, err := generateService(params)`?

Reply:
yes , line 336 in changed file @rikatz suggested to rename it to `service`

Reply:
Ah gotcha, so @rikatz is suggesting doing something similar to this https://github.com/kubernetes/kubernetes/blob/0ed8ee6fd7169219fbeb9bca83023a980a7a36a5/staging/src/k8s.io/kubectl/pkg/cmd/create/create_secret.go#L263

If you look at the code where `createSecret` will be called, the Secret created will then be put in this function https://github.com/kubernetes/kubernetes/blob/0ed8ee6fd7169219fbeb9bca83023a980a7a36a5/staging/src/k8s.io/kubectl/pkg/cmd/create/create_secret.go#L236. If you look at the function https://github.com/kubernetes/kubernetes/blob/0ed8ee6fd7169219fbeb9bca83023a980a7a36a5/staging/src/k8s.io/kubectl/pkg/util/apply.go#L141-L146 it will take Secret Object as an argument even though the type is not the same. It is because just like I mentioned `runtime.Object` is an interface (https://github.com/kubernetes/apimachinery/blob/v0.20.4/pkg/runtime/interfaces.go#L299).

So what you would want to do is to create a function called 
```
func (o *ExposeServiceOptions) createService() (*corev1.Service, error) 
```
and call  `service, err := o.CreateService()` over here
https://github.com/kubernetes/kubernetes/blob/0ed8ee6fd7169219fbeb9bca83023a980a7a36a5/staging/src/k8s.io/kubectl/pkg/cmd/expose/expose.go#L316 and treat the Service Object as runtime.Object.

Reply:
yes, exactly :D

Reply:
> yes, exactly :D
> 
> My idea here is that we turn this more a specific command (and maybe tightly related to kubectl create service) instead of using runtime.Objects here :)

There seems to be some issue with this one as test is failing on it

Reply:
It's not the issue of `runtime.Object` it's the issue of [type](https://en.wikipedia.org/wiki/Type_system) assignment in these line of code.
```
		service, err := o.createService()
		if err != nil {
			return err
		}

		if inline := cmdutil.GetFlagString(cmd, "overrides"); len(inline) > 0 {
			codec := runtime.NewCodec(scheme.DefaultJSONEncoder(), scheme.Codecs.UniversalDecoder(scheme.Scheme.PrioritizedVersionsAllGroups()...))
			service, err = cmdutil.Merge(codec, service, inline)
			if err != nil {
				return err
			}
		}
```
when you create a service , the type is `*corev1.Service`, but when you do `service, err = cmdutil.Merge(codec, service, inline)` the `cmdutil.Merge` is returning `runtime.Obj` type. 
A simple example will make this clearer
package main
https://play.golang.org/p/-ZwayR06bXp

simply changing it to 
 ```
service, err := cmdutil.Merge(codec, service, inline)
```
This will tell the compiler to treat the `service` as a new variable which has type `runtime.Object`

Reply:
`Merge` here will take `runtime.Obj` but service will be`*corev1.Service`

Reply:
Yea, but as I mentioned [here](https://github.com/kubernetes/kubernetes/pull/99891#discussion_r593789346) `runtime.Object` type is an `Interface` , and [Service](https://github.com/kubernetes/kubernetes/blob/b559434c02f903dbcd46ee7d6c78b216d3f0aca0/staging/src/k8s.io/api/core/v1/types.go#L4390) is an object that implement the type `Interface`

some example can be found here (https://gobyexample.com/interfaces) , see how `rect` and `circle` are implemeneted

Reply:
Try play around with it and modify the codes below , run it , and see what happens
```
func foo(bar geometry) {
	fmt.Println(bar.area())
}

func main() {
	r := rect{width: 3, height: 4}
	c := circle{radius: 5}

	// The `circle` and `rect` struct types both
	// implement the `geometry` interface so we can use
	// instances of
	// these structs as arguments to `measure`.
	measure(r)
	measure(c)
	
	foo(r)
}

```

Reply:
Thanks for doing this but you don't need to go too basic, I know about these concepts. 
I have taken your https://github.com/kubernetes/kubernetes/pull/99891#discussion_r593789346 into consideration and will tell you if I have further doubts